### DISCUSSION TITLE
Multicomponent Linear Solve

### DISCUSSION THREAD

**Discussion by alahyari** [AMReX-Codes/amrex#2976] [Q&A] [answered]
Hello AMReX folks,

I have a question wrt the MC linear solve feature. I need to solve three Poisson equations in the form of $$\Delta u_i = f_i$$ one for each velocity component in three dimensions. All components are cell-centered.

What I do now is that I solve one Poisson equation independently for each component in a for loop. I was thinking to use the MC feature. My first question is that is there going to be any saving in computational time if I use the MC feature compared to what I have now? And my second question is that is there any sample implementation of MC linear solve for cell-centered data that I can refer to?

Thank you.

**Comment by WeiqunZhang** [comment, marked as answer]
Yes, I believe there will be saving if all three components are solved together.  Although our cell-centered Poisson only supports one component, we do have another solver that can be used for this.  In `Src/LinearSolvers/MLMG/`, there is `amrex::MLALaplacian` that can be used to solve $a A u - b \nabla u = f$.  Here $a$ and $b$ are scalars (i.e., `amrex::Real`) and $A$ is a cell-centered `MultiFab`.  If you use `a_ncomp=3` in the constructor of `MLALaplacian` and call `MLALaplacian::setScalars(0.0, -1.0)`, the system will become three Poisson equations.  Note that you don't need to call `MLALaplacian::setACoeffs` because $a$ is zero.  Alternatively, you can extend `amrex::MLPoisson` to support multiple components.  It should be straightforward and it should have better performance than using `MLALaplacian`, but my guess is probably not that much.
